mds: make threshold for MDS_TRIM configurable

emmericp opened the pull request:
The MDS_TRIM health warning currently triggers on a hard-coded factor 2 threshold, I've got a setup here that runs some very spiky workload where this health warning triggers daily. I don't want to adjust `mds_log_max_segments` further as the current value is fine 99.9% of the time; I'd just like to get rid of the warning when that spike hits

https://tracker.ceph.com/issues/45906

<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [x] References tracker ticket
- [x] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

emmericp commented:
jenkins render docs

ceph-jenkins commented:
Doc render available at http://docs.ceph.com/ceph-prs/35415/

batrick commented on src/common/legacy_config_opts.h:
```diff
@@ -426,6 +426,7 @@ OPTION(mds_log_max_events, OPT_INT)
 OPTION(mds_log_events_per_segment, OPT_INT)
 OPTION(mds_log_segment_size, OPT_INT)  // segment size for mds log, default to default file_layout_t
 OPTION(mds_log_max_segments, OPT_U32)
+OPTION(mds_log_warn_factor, OPT_FLOAT)
```
We don't add configs to this file anymore. Please remove.

batrick commented on src/common/options.cc:
```diff
@@ -7952,6 +7952,10 @@ std::vector<Option> get_mds_options() {
     .set_description("maximum number of segments which may be untrimmed"),
 
+    Option("mds_log_warn_factor", Option::TYPE_FLOAT, Option::LEVEL_ADVANCED)
+    .set_default(2.0)
```
`.set_flag(Option::FLAG_RUNTIME)`

batrick replied:
`.set_min(1.0)`

emmericp commented:
Sorry, i'm confused about options: Last time I I added n option I needed to add it to both legacy options and the normal options.

Your review mentioned to remove it from the legacy options but that causes the compile to fail (just like i remember it from last time I added an option?)

Can you point me to some documentation how to add an option without touching legacy_config_opts?

batrick commented on src/mds/Beacon.cc:
```diff
@@ -304,9 +304,9 @@ void Beacon::notify_health(MDSRank const *mds)
+  // Indicates MDS is not trimming promptly
   {
-    if (mds->mdlog->get_num_segments() > (size_t)(g_conf()->mds_log_max_segments * 2)) {
+    if (mds->mdlog->get_num_segments() > (size_t)(g_conf()->mds_log_max_segments * g_conf()->mds_log_warn_factor)) {
```
This `g_conf()->mds_log_warn_factor` access only works if the option exists in legacy_options.h. This was for backwards compatibility with the option infrastructure was updated.

You need to instead use the new access method: ` g_conf().get_val<double>("mds_log_warn_factor")`. Then it should compile.

Whoever told you to modify `legacy_options.h` was wrong, sorry.

emmericp commented:
jenkins test make check

emmericp commented:
test failure is completely unrelated (tried twice):

```
/home/jenkins-build/build/workspace/ceph-pull-requests/build/bin/unittest_rgw_dmclock_scheduler: error while loading shared libraries: libboost_thread.so.1.73.0: cannot open shared object file: No such file or directory
```

so I guess my code is okay

batrick commented:
jenkins test make check

batrick commented:
jenkins test make check
